crates/proof-of-sql/src/sql/proof_plans/union_exec.rs

```diff
@@ -0,0 +1,319 @@
+}
+
+impl ProverEvaluate for UnionExec {
+    #[tracing::instrument(name = "UnionExec::first_round_evaluate", level = "debug", skip_all)]
+    fn first_round_evaluate<'a, S: Scalar>(
+        &self,
+        builder: &mut FirstRoundBuilder,
+        alloc: &'a Bump,
+        table_map: &IndexMap<TableRef, Table<'a, S>>,
+    ) -> Table<'a, S> {
+        let inputs = self
+            .inputs
+            .iter()
+            .map(|input| input.first_round_evaluate(builder, alloc, table_map))
+            .collect::<Vec<_>>();
+        let res = table_union(&inputs, alloc, self.schema.clone()).expect("Failed to union tables");
+        builder.request_post_result_challenges(2);
+        builder.produce_one_evaluation_length(res.num_rows());
+        res
+    }
+
+    #[tracing::instrument(name = "UnionExec::prover_evaluate", level = "debug", skip_all)]
+    #[allow(unused_variables)]
+    fn final_round_evaluate<'a, S: Scalar>(
+        &self,
+        builder: &mut FinalRoundBuilder<'a, S>,
+        alloc: &'a Bump,
+        table_map: &IndexMap<TableRef, Table<'a, S>>,
+    ) -> Table<'a, S> {
+        let inputs = self
+            .inputs
+            .iter()
+            .map(|input| input.final_round_evaluate(builder, alloc, table_map))
+            .collect::<Vec<_>>();
+        let input_lengths = inputs.iter().map(Table::num_rows).collect::<Vec<_>>();
+        let res = table_union(&inputs, alloc, self.schema.clone()).expect("Failed to union tables");
+        let alpha = builder.consume_post_result_challenge();
+        let beta = builder.consume_post_result_challenge();
+        let input_columns: Vec<Vec<Column<'a, S>>> = inputs
+            .iter()
+            .map(|table| table.columns().copied().collect::<Vec<_>>())
+            .collect::<Vec<_>>();
+        let output_columns: Vec<Column<'a, S>> = res.columns().copied().collect::<Vec<_>>();
+        // Produce intermediate MLEs for the union
+        output_columns.iter().copied().for_each(|column| {
+            builder.produce_intermediate_mle(column);
+        });
+        // Produce the proof for the union
+        prove_union(
+            builder,
+            alloc,
+            alpha,
+            beta,
+            &input_columns,
+            &output_columns,
+            &input_lengths,
+            res.num_rows(),
+        );
+        res
+    }
+}
+
+/// Verifies the union of tables.
+///
+/// # Panics
+/// Should never panic if the code is correct.
+#[allow(clippy::too_many_arguments)]
+fn verify_union<S: Scalar>(
+    builder: &mut VerificationBuilder<S>,
+    alpha: S,
+    beta: S,
+    input_evals: &[&[S]],
+    output_eval: &[S],
+    input_one_evals: &[S],
+    output_one_eval: S,
+) {
+    assert_eq!(input_evals.len(), input_one_evals.len());
+    let c_star_evals = input_evals
+        .iter()
+        .zip(input_one_evals)
+        .map(|(&input_eval, &input_one_eval)| {
+            let c_fold_eval = alpha * input_one_eval + fold_vals(beta, input_eval);
```

Comment:
We want to ensure that c_star_eval is actually padded with zeros. Having input_one_eval here means that it can be padded with anything the prover wants, which is not secure.
```suggestion
            let c_fold_eval = alpha + fold_vals(beta, input_eval);
```